Native Command Error Handling (#277)
PowerShell/PowerShell-RFC: 1-Draft/RFC00XX-Native-Command-Error-Handling.md

     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find path '/Users/jasonhelmick/natcmdbash/nofile' because it does not exist.

returns False
With set -o equiv. $ErrorActionPreference = 'Stop' : should return non-zero (false)
Get-Content: /Users/jasonhelmick/natcmdbash/testfile.ps1:6
Line |
   6 |  Get-Content -path ./nofile | Write-Output "pipe statement after failu …
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find path '/Users/jasonhelmick/natcmdbash/nofile' because it does not exist.
```

> [!NOTE] A common configuration command for POSIX shells `set -euo pipefail` includes the `set -u`
> configuration which returns an error if any variable has not been previously defined. This is
> equivalent to the existing PowerShell `Set-StrictMode` and does not need to be addressed in this
> RFC.

### $PSNativeCommandUseErrorActionPreference

The `$PSNativeCommandUseErrorActionPreference` preference variable will be implemented as a boolean.

- When enabled (1), `$ErrorActionPreference` will affect native commands with described behavior.
- When disabled (0), `$ErrorActionPreference` will have no effect to the original behavior.

> - The default value is disabled (0) to maintain backward compatibility.

Reviewer: ```suggestion
- The default value is 0 (disabled) to maintain backward compatibility.
```

- For non-zero exit codes and except for the value `Ignore`, an `ErrorRecord` will be added to
  `$Error` that wraps the exit code and the command executed that returned the exit code.

### Error object

The reported error record object will be the new type: `NativeCommandException` with the following details:

| Property        | Definition
----------------  | -------------------
| ExitCode:       |  The exit code of the failed command.

> | ErrorID:        | `"Program {0} ended with non-zero exit code {1}"`, with the command name and the exit code, from resource string `ProgramFailedToComplete`.

Reviewer 2: One thing that concerns me is that we're kind of all over the place on terminology:  NativeCommand, Program and the official get-command type is Application.  That said, I'm not sure I Application by itself e.g. `Application "whoami.exe" ended with non-zero exit code 1`.  But to say `Native application` is kind of redundant.

| ErrorCategory:  | `ErrorCategory.NotSpecified`.
| object:         | Exit code
| Source:         | The full path to the application

> | ProcessInfo     | Details of failed command including path, exit code, and PID

Reviewer 2: I think it would be helpful to see the arguments a a string.  Or maybe rather than command path, show the "command line".  Thoughts?

Reviewer 2: Is the process ID of any use given that the process no longer exists?

Reviewer: > I think it would be helpful to see the arguments a a string. Or maybe rather than command path, show the "command line". Thoughts?

This is not guaranteed after terminating of the process - an attempt to read a property of the Process class is likely to raise an exception. We can report reliably only current script line string.

> Is the process ID of any use given that the process no longer exists?

This can be useful as correlation data with event logs.

Reviewer 2: After talking this over with wg-engine group we decided that process id is useful for log correlation and arguments should not be recorded because it make contain passwords / api tokens.

PowerShell. This is not part of this proposal, but added for clarity.

```bash
bash-3.2$ cat file
#!/bin/bash
/bin/echo "Without set -u : No output - No error is produced"
/bin/echo "$firstname"
/bin/echo ""
/bin/echo "With set -u : Equivalent to Set-StrictMode -Version 2.0"
set -u
/bin/echo "$firstname"
```

```output
bash-3.2$ file
Without set -u : No output - No error is produced

With set -u : Equivalent to Set-StrictMode -Version 2.0
/Users/jasonhelmick/natcmdbash/strict: line 10: firstname: unbound variable
```

```powershell
PS> cat ./file.ps1
/bin/echo "Without Set-StrictMode -version 2.0 : No output - No error is produced"
/bin/echo "$firstname"
/bin/echo ""
/bin/echo "With Set-StrictMode -version 2.0 : Equivalent to set -u"
Set-StrictMode -version 2.0
/bin/echo "$firstname"
```

```output
PS> ./file.ps1
Without set -u : No output - No error is produced

With Set-StrictMode -version 2.0 : Equivalent to set -u
InvalidOperation: /Users/jasonhelmick/natcmdbash/psstrict.ps1:9
Line |
   9 |  /bin/echo "$firstname"
     |             ~~~~~~~~~~
     | The variable '$firstname' cannot be retrieved because it has not been set.
```

### set -e/ $ErrorActionPreference

> In the example below, `set -e` is not equivalent to `$ErrorActionPReference` for native commands.

Reviewer 3: ```suggestion
In the example below, `set -e` is not equivalent to `$ErrorActionPreference` for native commands.
```

```bash
bash-3.2$ cat file
#!/bin/bash
/bin/echo "Without set -e : Will receive message after failure"
/bin/cat ./nofile
/bin/echo "Message After failure"
/bin/echo ""

> /bin/echo "With set -e : Will NOT receive message after failure"

Reviewer 3: should say "Will NOT continue script execution after failure"

/bin/echo "With `$ErrorActionPreference = 'Stop' : SHOULD NOT receive message after failure - but does"
$ErrorActionPreference = "Stop"
/bin/cat ./nofile
/bin/echo "Message After failure"
/bin/echo ""
Write-Host "With cmdlet's - `$ErrorActionPreference = 'Stop' : Will NOT receive message after failure"
$ErrorActionPreference = 'Stop'
Get-Content ./nofile
Write-Host "Message After failure"
```

```output
PS> ./file.ps1
Without $ErrorActionPreference : Will receive message after failure
cat: ./nofile: No such file or directory
Message After failure

With $ErrorActionPreference = 'Stop' : SHOULD NOT receive message after failure - but does
cat: ./nofile: No such file or directory
Message After failure

With cmdlet's - $ErrorActionPreference = 'Stop' : Will NOT receive message after failure
Get-Content: /Users/jasonhelmick/natcmdbash/psstop.ps1:17
Line |
  17 |  Get-Content ./nofile
     |  ~~~~~~~~~~~~~~~~~~~~
     | Cannot find path '/Users/jasonhelmick/natcmdbash/nofile' because it does not exist.
```

### set -o pipefail/ PSNativeCommandErrorAction

In the example below, PowerShell has no equivalent to `set -o pipefail` for native commands.

```bash
bash-3.2$ cat file
#!/bin/bash

> /bin/echo "Without set -o pipefail : returns 0"

Reviewer 3: Need to clarify what 0 and 1 means for those not familiar

| Continue        | Displays the error message and continues executing.
| Ignore          | Suppresses the error message and continues to execute the command.
| Inquire         | Displays the error message and asks you whether you want to continue.
| SilentlyContinue| The error message isn't displayed and execution continues without interruption.
| Stop            | Displays the error message and stops executing. In addition to the error generated, the Stop value generates an ActionPreferenceStopException object to the error stream. stream
| Suspend         | Automatically suspends a job to allow for further investigation.

### Preference variable resolves error handling conflicts

In cases where an existing script already handles non-zero native command errors, the preference
variable `$PSNativeCommandErrorAction` may be set to `Continue`. Error handling behavior of the
PowerShell cmdlets is handled separately with `$ErrorActionPreference`.

### Error object

The reported error record object will be the new type: `NativeCommandException` with the following details:

| Property        | Definition
----------------  | -------------------

> | ExitCode:      |  The exit code of the failed command.

Reviewer 3: some alignment issues with vertical bars

occurs. This is particularly true in CI where the preference is to fail fast. PowerShell currently
supports customers with this ability by setting the `$ErrorActionPreference` variable in the script
to `Stop`.

```Powershell
$ErrorActionPreference = 'Stop'
```

Native commands usually return an exit code to the calling application which will be zero for
success or non-zero for failure. However, native commands currently do not participate in the
PowerShell error stream. Redirected `stderr` output is not interpreted the same as the PowerShell
error stream as many native commands use `stderr` as information/verbose stream and thus only the
exit code matters. Users working with native commands in their scripts will need to check the
execution status after each call using a helper function similar to below:

```Powershell
if ($LASTEXITCODE -ne 0)
{
    throw "Command failed. See above errors for details"
}
 ```

Simply relaying the errors through the error stream isn't the solution. The example itself doesn't
support all cases as `$?` can be false from a cmdlet or function error, making `$LASTEXITCODE`
stale.

> In POSIX shells, this need to terminate on command error is addressed by the `set -e` configuration,

Reviewer 4: ```suggestion
In POSIX shells, terminating execution when a command has an error is enabled via executing `set -e` in the session.
```

> which causes the script to exit when a command fails. In addition, to ensure that an error is

Reviewer 4: ```suggestion
In addition, to ensure that an error is
```

> returned if any command in a pipeline fails, POSIX shells address this need with `set -o pipefail`

Reviewer 4: ```suggestion
returned if any command in a pipeline fails, POSIX shells address this via executing `set -o pipefail`
```

> configuration.

Reviewer 4: ```suggestion
in the session.
```

This specification proposes a similar idea, but adapted to the PowerShell conventions of preference
variables and catchable, self-describing, terminating error objects. This proposal adds the

> equivalent functionality of `set -eo pipefail` to stop execution and return an error if any command in a pipeline

Reviewer 4: ```suggestion
equivalent functionality of `set -e -o pipefail` (abbreviated to `set -eo pipefail`) to stop execution and return an error if any command in a pipeline
```

fails.

 The specification and alternative proposals are based on the
 [Equivalent of bash `set -e` #3415](https://github.com/PowerShell/PowerShell/issues/3415)
 committee review of the associated
 [pull request](https://github.com/PowerShell/PowerShell/pull/3523), and
 [implementation plan](https://github.com/PowerShell/PowerShell-RFC/pull/88#issuecomment-613653678)

## Specification

> This RFC proposes a preference variable to configure the elevation of errors produced by native

Reviewer 4: ```suggestion
This RFC proposes a preference variable to enable errors produced by native
```

> commands to first-class PowerShell errors, so that native command failures will produce error

Reviewer 4: ```suggestion
commands to be PowerShell errors, so that failures will produce error
```

objects that are added to the error stream and may terminate execution of the script without added
boilerplate.

The specification proposes similar functionality to the common POSIX shell configuration `set -eo pipefail`.

> - `set -u` - returns an error if any variable has not been previously defined.

Reviewer 4: this isn't really part of the new proposal, right?

Author: Correct, I'll add a sentence that says this was added for clarity.

- `set -e` - instructs to immediately exit if any command has a non-zero exit status.
- `set -o pipefail` - prevents errors in the pipeline from being masked. The return code for the
  non-zero error is returned for the entire pipeline.

### set -u/ Set-StrictMode

> In the example below, `set -u` is equivalent to `Set-StrictMode -Version 2.0`.

Reviewer 4: ```suggestion
In the example below, `set -u` behavior of `bash` is shown followed by the proposed behavior for PowerShell.
```

> 

Reviewer 4: i might do this a little differently, just to reduce confusion
```sh
bash-3.2$ cat showerr.sh
#!/bin/bash
/bin/echo "Without set -e : Will receive message after failure"
/bin/cat ./nofile
/bin/echo "Message After failure"
/bin/echo ""
/bin/echo "With set -e : Will NOT receive message after failure"
set -e
/bin/cat ./nofile
/bin/echo "Message After failure"

bash-3.2$ ./showerr.sh
Without set -e : Will receive message after failure
cat: ./nofile: No such file or directory
Message After failure

With set -e : Will NOT receive message after failure
cat: ./nofile: No such file or directory
bash-3.2$ 
```

Reviewer 4: then provide the powershell equivalent

> equivalent to the existing PowerShell `Set-StrictMode` and does not need to be addressed in this
> RFC.

### $PSNativeCommandErrorAction

The `$PSNativeCommandErrorAction` preference variable will implement a version of the
`$ErrorActionPreference` variable for native commands.

- The value will default to `Continue` for compatibility with existing behavior.
- For non-zero exit codes and except for the value `Ignore`, an `ErrorRecord` will be added to
  `$Error` that wraps the exit code and the command executed that returned the exit code.
- Initially, only the existing values of `$ErrorActionPreference` will be supported in
  `$PSNativeCommandErrorAction` as described in the table below.

### Valid values for `$PSNativeCommandErrorAction`

| Value           | Definition
----------------  | -------------------
| Break           | Enter the debugger when an error occurs or when an exception is raised.
| Continue        | Displays the error message and continues executing.
| Ignore          | Suppresses the error message and continues to execute the command.
| Inquire         | Displays the error message and asks you whether you want to continue.

> | SilentlyContinue| No effect. The error message isn't displayed and execution continues without interruption.

Reviewer 4: this really is an effect, right?
```suggestion
| SilentlyContinue| The error message isn't displayed and execution continues without interruption.
```

| Stop            | Displays the error message and stops executing. In addition to the error generated, the Stop value generates an ActionPreferenceStopException object to the error stream. stream

> | Suspend         | Automatically suspends a workflow job to allow for further investigation.

Reviewer 4: hmmm, workflow? no workflow in PS7 - do you mean job?

### Preference variable resolves error handling conflicts

In cases where an existing script already handles non-zero native command errors, the preference
variable `$PSNativeCommandErrorAction` may be set to `Continue`. Error handling behavior of the
PowerShell cmdlets is handled separately with `$ErrorActionPreference`.

### Error object

The reported error record object will be the new type: `NativeCommandException` with the following details:

| Property        | Definition
----------------  | -------------------
| ExitCode:      |  The exit code of the failed command.
| ErrorID:       | `"Program {0} ended with non-zero exit code {1}"`, with the command name and the exit code, from resource string `ProgramFailedToComplete`.
| ErrorCategory: | `ErrorCategory.NotSpecified`.

> | object:         | exit code

Reviewer 4: TargetObject?

| Source:        | The full path to the application
| ProcessInfo    | details of failed command including path, exit code, and PID

## Alternative Approaches and Considerations

### Native commands should respect $ErrorActionPreference

> Native commands should respect $ErrorActionPreference and not need $PSNativeCommandErrorAction. This

Reviewer 4: ```suggestion
Native commands should use the $ErrorActionPreference setting and not need $PSNativeCommandErrorAction. This
```

could be released as an experimental feature.

### Extending $PSNativeCommandErrorAction

For users that prefer to set a single preference variable that affects both PowerShell and native
commands, The values of `$PSNativeCommandErrorAction` may be extended to include
`MatchErrorActionPreference`, which should apply the `$ErrorActionPreference` setting to native
commands.

> - A conversion will occur between `$PSNativeCommandErrorAction` and `$ErrorActionPreference` values,

Reviewer 4: If `$PSNativeCommandErrorAction` is set to `MatchErrorActionPreference` then the value of `$ErrorActionPreference` will define the behavior for native command errors.

### Preference variable resolves error handling conflicts

In cases where an existing script already handles non-zero native command errors, the preference

> variable `$PSNativeCommandUseErrorActionPreference` may be set to `$false`. Error handling behavior

Reviewer 2: Making this name longer seems like the wrong direction to go.  Also, if it ends in `ActionPreference` then I would expect it to be of type `[System.Management.Automation.ActionPreference]` instead of `[bool]`.

Reviewer 5: I agree that the variable being boolean and the name ending with `Preference` sounds weird. `PSNativeCommandErrorActionEnabled`,  `PSEnableNativeCommandErrorAction` better describes the implicit type of the variable. Still they are long names, though.
